### PR TITLE
Add support for correctly discovering async test methods on non-Apple platforms

### PR DESCRIPTION
**Pull request by briancroom**
This adopts [a recent enhancement to swift-corelibs-xctest for supporting `async` test methods](https://github.com/apple/swift-corelibs-xctest/pull/331) by ensuring the automatic test discovery machinery used on non-Apple platforms generates the correct test manifest code for the harness to run async tests.

### Modifications:

I have adjusted the code-gen logic used during test discovery to notice async test methods, using the richer IndexStore data provided by https://github.com/apple/swift-tools-support-core/pull/258, and wrap such methods in the `asyncTest` helper from swift-corelibs-xctest so that they can execute correctly.

### Result:

With this change, it is now possible to transparently run async XCTests on Linux via `swift test`.

### CODE REVIEW

**Comment by neonichu** [comment]
@swift-ci please smoke test

**Review by neonichu — APPROVED** [review]
Thanks, Brian!

**Comment by neonichu** [comment]
https://github.com/apple/swift-tools-support-core/pull/258
@swift-ci please smoke test

**Comment by briancroom** [comment]
Note that I haven't seen the new test pass on Linux yet. I tried using one of the "nightly" Docker images, but they're about a week out of date right now and don't have the swift-corelibs-xctest change in place yet. We'll see how CI goes.

File: Fixtures/Miscellaneous/TestDiscovery/Async/Tests/AsyncTests/SwiftTests.swift [hunk 1]
```diff
@@ -0,0 +1,17 @@
+import XCTest
+@testable import Async
+
+class AsyncTests: XCTestCase {
+
+    func testAsync() async {
+        XCTAssertFalse(Thread.isMainThread)
```
Comment (stmontgomery) [inline Fixtures/Miscellaneous/TestDiscovery/Async/Tests/AsyncTests/SwiftTests.swift:7]: Are we confident that an `async`, non-`@MainActor` func is _guaranteed_ to never run on the main thread, even by coincidence? I bet that is true currently and perhaps always will be, although I feel like semantically there is no particular guarantee there

**Review by stmontgomery — APPROVED** [review]
Looks great! I left a comment on the TSC PR which may impact this, but the changes here all look good to me

**Comment by neonichu** [comment]
apple/swift-tools-support-core#258
@swift-ci please smoke test

File: Fixtures/Miscellaneous/TestDiscovery/Async/Tests/AsyncTests/SwiftTests.swift (hunk 1 above)
Comment (briancroom) [reply Fixtures/Miscellaneous/TestDiscovery/Async/Tests/AsyncTests/SwiftTests.swift]: Yeah, that's fair. The assertion isn't really providing significant value anyway, so I'll just remove it.

**Comment by briancroom** [comment]
apple/swift-tools-support-core#258
@swift-ci please smoke test

**Comment by briancroom** [comment]
@swift-ci please smoke test

**Comment by briancroom** [comment]
It looks like the “Linux (self hosted)” CI job uses the latest pre-built trunk development toolchain, which unfortunately is a week out of date right now, and is missing the swift-corelibs-xctest changes needed for the integration test to succeed, so we’re effectively blocked on landing this until a newer nightly toolchain can get posted (cc @shahmishal, who I spoke to about this earlier.)

**Comment by neonichu** [comment]
The self-hosted jobs do not gate merging.

**Comment by briancroom** [comment]
> The self-hosted jobs do not gate merging.

Ah ok, good to know. So you wouldn’t be concerned if this goes in before the toolchain updates to allow that to pass? (Assuming the other, required, Linux job does indeed pass now?)

**Comment by neonichu** [comment]
> So you wouldn’t be concerned if this goes in before the toolchain updates to allow that to pass? 

Nope, the main reason for the self hosted jobs is to have quicker iteration time than the smoke tests which take hours, they're not really testing anything in addition to the smoke tests. So I think this is good to merge.

**Comment by briancroom** [comment]
Excellent! The primary Linux job passed, and I confirmed in the logs that it indeed ran the new test. That gives me enough confidence to proceed with landing this. Thanks everybody! 🎉

**Comment by tomerd** [comment]
thanks @briancroom

**Pull request merged**